### ISSUE TITLE
UPnP not Sending Audio / No Sound

### ISSUE THREAD

**Issue by radoprajitno** [music-assistant/hass-music-assistant#1829]
### What version of Music Assistant has the issue?

 2024.1.0

### The problem

I'm having issues playing audio via UPnP from Music Assistant.
My device is Sony TV X85J & Bubble UPNP on Xiaomi Pad 6

- I am able to cast to my TV via Chromecast from Music Assistant
- NOT able to cast to my TV via UPnP from Music Assistant (I'm able to control the volume & see media info playing on my TV)

However:
- I am able to cast via UPnP to my TV via Home Assistant Media Browser

I can't find related logs that show failure / error during the UPnP transmission.

I'm Running the latest version of Home Assistant OS
Core 2024.1.5
Supervisor 2023.12.1
Operating System 11.4

I'm using Music Assistant via HACS Installation method for the server and integration
Server 2.0.0b85
Integration 2024.1.0

### How to reproduce

-

### What version of Home Assistant Core are your running

 2024.1.0

### What type of installation are you running?

Home Assistant OS

### On what type of hardware are you running?

Alternative

**Comment by OzGav** [comment]
DLNA/UPnP isn't a great protocol. If Chromecast works then just use that.

You mention DLNA works from the HA media browser but which entity are you targetting. The HA entitity or the MA entity (which will only be visible if you have installed the integration)

**Comment by radoprajitno** [comment]
The reason im sticking to UPnP is to send better quality in terms of Audio Quality to my Sony TV.

Chromecast has a limit cap on transfering Audio Quality.

I does sound better.

**Comment by OzGav** [comment]
Do you have any supporting logs?

**Comment by radoprajitno** [comment]
Meanwhile my TV is at the other room,
I have tried to cast to a BubbleUPnP (Android Apps), and managed to find logs that might related.

Still, unable to cast to it (No Audio), yet i can control the volume of the devices.

Here it is
 
2024-01-25 21:48:39.473 DEBUG (MainThread) [async_upnp_client.client] Calling action: GetTransportInfo, args: {'InstanceID': 0}
2024-01-25 21:48:39.473 DEBUG (MainThread) [async_upnp_client.traffic.upnp] Sending request:
POST http://192.168.1.199:58645/dev/92b44e83-7e27-4c89-bef0-9aa9572423e4/svc/upnp-org/AVTransport/action
SOAPAction: "urn:schemas-upnp-org:service:AVTransport:1#GetTransportInfo"
Host: 192.168.1.199:58645
Content-Type: text/xml; charset="utf-8"
<?xml version="1.0"?><s:Envelope s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Body><u:GetTransportInfo xmlns:u="urn:schemas-upnp-org:service:AVTransport:1"><InstanceID>0</InstanceID></u:GetTransportInfo></s:Body></s:Envelope>

2024-01-25 21:48:39.498 DEBUG (MainThread) [async_upnp_client.traffic.upnp] Got response from POST http://192.168.1.199:58645/dev/92b44e83-7e27-4c89-bef0-9aa9572423e4/svc/upnp-org/AVTransport/action:
200
Ext: 
Server: Linux/4.4.302-perf UPnP/1.0 BubbleUPnP/4.3.2
Content-Type: text/xml;charset=utf-8
Connection: close
Date: Thu, 25 Jan 2024 14:48:39 GMT
Content-Length: 454

b'<?xml version="1.0" encoding="utf-8" standalone="yes"?><s:Envelope s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Body><u:GetTransportInfoResponse xmlns:u="urn:schemas-upnp-org:service:AVTransport:1"><CurrentTransportState>STOPPED</CurrentTransportState><CurrentTransportStatus>OK</CurrentTransportStatus><CurrentSpeed>1</CurrentSpeed></u:GetTransportInfoResponse></s:Body></s:Envelope>'
2024-01-25 21:48:39.498 DEBUG (MainThread) [async_upnp_client.client] Called action: GetTransportInfo, args: {'InstanceID': 0}, response_args: {'CurrentTransportState': 'STOPPED', 'CurrentTransportStatus': 'OK', 'CurrentSpeed': '1'}
2024-01-25 21:48:40.163 DEBUG (MainThread) [music_assistant.providers.dlna] Handling command play_media for player Mi 6 UPnP
2024-01-25 21:48:40.164 DEBUG (MainThread) [async_upnp_client.profiles.dlna] Set transport uri: http://192.168.1.7:8096/uuid:92b44e83-7e27-4c89-bef0-9aa9572423e4/single/6c2308b69f9b45e9909e87c1315387fc.flac?ts=1706194120
2024-01-25 21:48:40.164 DEBUG (MainThread) [async_upnp_client.client] Calling action: SetAVTransportURI, args: {'InstanceID': 0, 'CurrentURI': 'http://192.168.1.7:8096/uuid:92b44e83-7e27-4c89-bef0-9aa9572423e4/single/6c2308b69f9b45e9909e87c1315387fc.flac?ts=1706194120', 'CurrentURIMetaData': '<DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/" xmlns:dlna="urn:schemas-dlna-org:metadata-1-0/"><item id="1" parentID="0" restricted="1"><dc:title>Easy On Me</dc:title><dc:creator>Adele</dc:creator><upnp:album>30</upnp:album><upnp:artist>Adele</upnp:artist><upnp:duration>225</upnp:duration><dc:queueItemId>6c2308b69f9b45e9909e87c1315387fc</dc:queueItemId><upnp:albumArtURI>https://resources.tidal.com/images/a173ee14/db90/4f21/b530/0f7bab7d95e2/640x640.jpg</upnp:albumArtURI><upnp:class>object.item.audioItem.audioBroadcast</upnp:class><upnp:mimeType>audio/flac</upnp:mimeType><res duration="0:03:45.000" protocolInfo="http-get:*:audio/flac:DLNA.ORG_PN=FLAC;DLNA.ORG_OP=01;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=0d500000000000000000000000000000">http://192.168.1.7:8096/uuid:92b44e83-7e27-4c89-bef0-9aa9572423e4/single/6c2308b69f9b45e9909e87c1315387fc.flac?ts=1706194120</res></item></DIDL-Lite>'}
2024-01-25 21:48:40.164 DEBUG (MainThread) [async_upnp_client.traffic.upnp] Sending request:
POST http://192.168.1.199:58645/dev/92b44e83-7e27-4c89-bef0-9aa9572423e4/svc/upnp-org/AVTransport/action
SOAPAction: "urn:schemas-upnp-org:service:AVTransport:1#SetAVTransportURI"
Host: 192.168.1.199:58645
Content-Type: text/xml; charset="utf-8"
<?xml version="1.0"?><s:Envelope s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Body><u:SetAVTransportURI xmlns:u="urn:schemas-upnp-org:service:AVTransport:1"><InstanceID>0</InstanceID>
<CurrentURI>http://192.168.1.7:8096/uuid:92b44e83-7e27-4c89-bef0-9aa9572423e4/single/6c2308b69f9b45e9909e87c1315387fc.flac?ts=1706194120</CurrentURI>
<CurrentURIMetaData>&lt;DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/" xmlns:dlna="urn:schemas-dlna-org:metadata-1-0/"&gt;&lt;item id="1" parentID="0" restricted="1"&gt;&lt;dc:title&gt;Easy On Me&lt;/dc:title&gt;&lt;dc:creator&gt;Adele&lt;/dc:creator&gt;&lt;upnp:album&gt;30&lt;/upnp:album&gt;&lt;upnp:artist&gt;Adele&lt;/upnp:artist&gt;&lt;upnp:duration&gt;225&lt;/upnp:duration&gt;&lt;dc:queueItemId&gt;6c2308b69f9b45e9909e87c1315387fc&lt;/dc:queueItemId&gt;&lt;upnp:albumArtURI&gt;https://resources.tidal.com/images/a173ee14/db90/4f21/b530/0f7bab7d95e2/640x640.jpg&lt;/upnp:albumArtURI&gt;&lt;upnp:class&gt;object.item.audioItem.audioBroadcast&lt;/upnp:class&gt;&lt;upnp:mimeType&gt;audio/flac&lt;/upnp:mimeType&gt;&lt;res duration="0:03:45.000" protocolInfo="http-get:*:audio/flac:DLNA.ORG_PN=FLAC;DLNA.ORG_OP=01;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=0d500000000000000000000000000000"&gt;http://192.168.1.7:8096/uuid:92b44e83-7e27-4c89-bef0-9aa9572423e4/single/6c2308b69f9b45e9909e87c1315387fc.flac?ts=1706194120&lt;/res&gt;&lt;/item&gt;&lt;/DIDL-Lite&gt;</CurrentURIMetaData></u:SetAVTransportURI></s:Body></s:Envelope>

**Comment by radoprajitno** [comment]
From HA Music Assistant Entity, playing can be done correctly,
And it seems it send a bit perfect Audio Quality, without any transcoding happening.

My rough guess::

Im runing HAOS on a proxmox VM, it doesnt have any sound card access.
Might it be MA DLNA/UPnP side is trying to transcode the Audio?

**Comment by marcelveldt** [comment]
> The reason im sticking to UPnP is to send better quality in terms of Audio Quality to my Sony TV.
> 
> Chromecast has a limit cap on transfering Audio Quality.
> 
> I does sound better.

You will really not be able to hear the difference...
Chromecast is capped to 96khz/24 bits HI-RES audio while DLNA theoretically supports 192/24 but it depends on ythe player. Nobody can hear the difference between (lossless!) 44.1/16 and 48/24, let alone any higher. If you can hear a difference, then you must see a doctor ;-)

What CAN actually be happening is that your device is hurting the audio, for example throwing away bots from an 24 bits source signal instead of dithering and whatnot. Consumer devices like TV's are not exactly meant to produce high quality audio so the firmware designers of those devices usually don't give a damn about audio quality, especially not hi res.

Chromecast on TV's is usually limited to the video version of chromecast (no hi res, 48/16 capped) while the audio version of Chromecast (built in to speakers and receivers) supports 96/24, so super high quality audio.

In your case you are saying that playing to the DLNA player doesn't work from MA.
Well, by default MA sends the lossless FLAC stream to the player but a lot of players can only accept MP3 (or other lossy codecs) on the DLNA interface. 

So long story short, your TV does not support lossless (FLAC) audio in the DLNA interface, you should use cast instead.
Or Airplay, which is also lossless (but limited to redbook 44.1/16 quality) but again, nobody can hear the difference between a proper lossless 44.1/16 bits stream versus any higher sample rates or bit depth, unless maybe on some 40K speakers some good trained ears will hear it, but certainly not on your TV.

You WILL be able to hear a difference between a lossy codec such as MP3 and lossless cd quality.

**Comment by radoprajitno** [comment]
> > The reason im sticking to UPnP is to send better quality in terms of Audio Quality to my Sony TV.
> > 
> > Chromecast has a limit cap on transfering Audio Quality.
> > 
> > I does sound better.
> 
> You will really not be able to hear the difference...
> Chromecast is capped to 96khz/24 bits HI-RES audio while DLNA theoretically supports 192/24 but it depends on ythe player. Nobody can hear the difference between (lossless!) 44.1/16 and 48/24, let alone any higher. If you can hear a difference, then you must see a doctor ;-)
> 
> What CAN actually be happening is that your device is hurting the audio, for example throwing away bots from an 24 bits source signal instead of dithering and whatnot. Consumer devices like TV's are not exactly meant to produce high quality audio so the firmware designers of those devices usually don't give a damn about audio quality, especially not hi res.
> 
> Chromecast on TV's is usually limited to the video version of chromecast (no hi res, 48/16 capped) while the audio version of Chromecast (built in to speakers and receivers) supports 96/24, so super high quality audio.
> 
> In your case you are saying that playing to the DLNA player doesn't work from MA.
> Well, by default MA sends the lossless FLAC stream to the player but a lot of players can only accept MP3 (or other lossy codecs) on the DLNA interface. 
> 
> So long story short, your TV does not support lossless (FLAC) audio in the DLNA interface, you should use cast instead.
> Or Airplay, which is also lossless (but limited to redbook 44.1/16 quality) but again, nobody can hear the difference between a proper lossless 44.1/16 bits stream versus any higher sample rates or bit depth, unless maybe on some 40K speakers some good trained ears will hear it, but certainly not on your TV.
> 
> You WILL be able to hear a difference between a lossy codec such as MP3 and lossless cd quality.
> 
> 
> 
> 


DLNA and UPnP is the backbone for HIFI enthusiast streamer. I believe "Tidal Connect" is based on UPnP with their own proprietary of additional layers.

In deed it is very hard to listen for a difference. We will need proper gear or audio setup.

For chromecast: ive found articles that it is able to stream at 24/96 Quality. However it is "Chromecast Audio" specific device.

Howeverr:::

While I'm experimenting a great way "streaming"to a android TV (Altenative or equivalent quality as Tidal Connect), 
Here are things that I"ve found:

MA DLNA is sending "Same as source" formats, either is losseless or lossy. All withouy any transcode happening.

I encountered the problem: because my device simply unable to decode / play the hires 24bit/48 and above. In this case all of my device os android based system. 

For an android TV, ive seen from my roon signal path, it is outputing at 16/44.1 (CD quality). And its capped at that quality).

However the game changes when i Plug my USB dac, and use UAPP (USB Audio Player Pro). It has features to bypass Internal Audio Driver and uses USB DAC Driver instead. It also has a UPnP streamer options.

With that setup, MA DLNA/UPnP works flawlessly.

My personal opinion:

MA has a huge potential to compete with other services such as Volumio, Roon, in the future.

I'd love to see how it progress.
